flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -155,6 +162,12 @@ private void monitorAndForwardSplits() {
       if (lastSnapshotId == INIT_LAST_SNAPSHOT_ID) {
         newScanContext = scanContext.copyWithSnapshotId(snapshotId);
       } else {
+        List<Long> snapshotIds = SnapshotUtil.snapshotIdsBetween(table, lastSnapshotId, snapshot.snapshotId());
+        if (snapshotIds.size() < scanContext.monitorSnapshotNumber()) {
+          snapshotId = snapshot.snapshotId();
+        } else {
+          snapshotId = snapshotIds.get(snapshotIds.size() - scanContext.monitorSnapshotNumber());
+        }
```

Comment:
Can you elaborate on this logic here / walk me through an example case where `snapshotId` needs to be determined because it's equal to (or possibly greater than?) the `monitorSnapshotNumber`?

Reply:
Instead, I suggest using `lastSnapshotId` and `monitorSnapshotNumber` directly to get to the destination.

Specifically, there should be two ways:

1.  use `lastSnapshotId` and `monitorSnapshotNumber` to get `snapshotId`, instead of getting `currentSnapshotId` and comparing,
2.  Use `lastSnapshotId` and `monitorSnapshotNumber` directly to get the snapshot list

The above may require modifications to the core module, but I think it will make the logic more intuitive.

Reply:
I’m thinking through these situations now.

But users would _never enter this block_ if they don’t opt into the new behavior, is that correct?

If so, can we add a conditional case that this block won’t be entered unless the user have a non `INT_MAX` value (eg we make a Precondition and just make a sentence that starts with `[bug] this shouldn’t happen <because>`). We have one or two other places that use the same `[bug]` syntax and this new logic change would ideally not apply to users that keep the default behavior.

Reply:
I do like the idea of using the snapshot ID though generally speaking (in this case I need to review).

However, It’s possible to turn off snapshot ID inheritance. So we’d need to consider that.

Reply:
Also, can we make this its own method?

Reply:
Since we already consider the inheritance in state initialization, do we need another check?

@hililiwei I agree computing using `lastComsumedSnapshotId` and `maxSnapshotsPerMonitorInterval` is more accurate. But I haven't found direct methods or utils to compute that, maybe need more utils in `SnapshotUtil` as you said. Can we have such cases in which the last consumed snapshot is not an ancestor of the latest snapshot?

Reply:
I wonder if that would be the case after supporting `Branch and Tag` feature?

Reply:
I'm not sure whether it can route to different branches without changing table states or properties. Like git, if we don't explicitly `checkout` a branch we should be able to traverse the history, right?